### PR TITLE
[INTEL MKL] Add support for mkl_quantize_op. fp32->{uint8, int8}

### PR DESCRIPTION
**Pull request by Srini511**
This PR adds support for MKL quantize operator to convert activations in  fp32 to 8 bits precision

### CODE REVIEW

**Comment by Srini511** [comment]
Hi @penpornk . Gentle reminder for review.

**Comment by Srini511** [comment]
@penpornk gentle reminder.

File: tensorflow/core/ops/mkl_array_ops.cc [hunk 1]
```diff
@@ -87,6 +87,34 @@ REGISTER_OP("_MklQuantizedConcatV2")
...
+      ShapeHandle unused;
+      TF_RETURN_IF_ERROR(c->WithRank(c->input(1), 0, &unused));
+      TF_RETURN_IF_ERROR(c->WithRank(c->input(2), 0, &unused));
+      c->set_output(1, c->Scalar());
+      c->set_output(2, c->Scalar());
+      return Status::OK();
+    });
 }  // namespace tensorflow
 
 #endif
```
Comment (penpornk) [inline tensorflow/core/ops/mkl_array_ops.cc]:
```suggestion
#endif  // INTEL_MKL
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you very much for the PR, and I'm so sorry for my delay!

File: tensorflow/core/graph/mkl_layout_pass.cc [hunk 2]
```diff
@@ -1276,6 +1281,31 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
...
+    if (mode_string != "SCALED" || round_mode_string != "HALF_TO_EVEN") {
+      VLOG(1) << "QuantizeOpRewrite: Mode is not SCALED and/or"
+              << "rounding mode is not HALF_TO_EVEN"
+              << "This case is not optimized by Intel MKL, thus using Eigen op"
+              << "for Quantize op ";
+      return false;
+    }
+    if (filter_node->IsConstant()) {
+      VLOG(1) << "QuantizeOpRewrite: Trying to quantize a node which "
+              << "could possibly be a filter"
```
[the remark this answers is not in the document]
Comment (Srini511) [reply tensorflow/core/graph/mkl_layout_pass.cc]: Agree

File: tensorflow/core/kernels/BUILD [hunk 3]
```diff
@@ -6444,6 +6444,58 @@ tf_cc_test(
     ],
 )
 
+tf_mkl_kernel_library(
+   name = "mkl_quantized_op",
+   srcs = [
```
[the remark this answers is not in the document]
Comment (Srini511) [reply tensorflow/core/kernels/BUILD]: Agree

File: tensorflow/core/kernels/BUILD [hunk 4]
```diff
@@ -6444,6 +6444,58 @@ tf_cc_test(
...
+       ":quantization_utils",
+       ":transpose_functor",
+       "//tensorflow/core:array_ops_op_lib",
+       "//tensorflow/core:core_cpu",
+       "//tensorflow/core:framework",
+       "//tensorflow/core:lib",
+       "//tensorflow/core:math_ops_op_lib",
+       "//tensorflow/core:nn_ops_op_lib",
+       "//third_party/eigen3",
+       "//third_party/mkl:intel_binary_blob",
```
[the remark this answers is not in the document]
Comment (Srini511) [reply tensorflow/core/kernels/BUILD]: agree

File: tensorflow/core/kernels/mkl_quantize_op.cc [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    // downstream ops when they need to do calculations on them.
+    // We pick the value by making sure that zero is not more than 100x the
+    // overall range from the maximum, so that the value can be easily
+    // represented when we promote the quantized value to a higher
+    // intermediate bit depth, since that's a common requirement.
+    const float epsilon = std::max(1.0f, std::max(fabsf(input_min_range),
+                                                  fabsf(input_max_range))) /
+                          100.0f;
+    max_range = std::max(input_max_range, min_range + epsilon);
+    max_range = std::max(0.0f, max_range);
```
[the remark this answers is not in the document]
Comment (Srini511) [reply tensorflow/core/kernels/mkl_quantize_op.cc]: we cannot say that if both min_range and max_range are both negative.

File: tensorflow/core/kernels/mkl_quantize_op.cc [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+    MklDnnShape output_mkl_shape;
+    TensorShape output_tf_shape;
+    if (src_mkl_shape.IsMklTensor()) {
+      output_mkl_shape.SetMklTensor(true);
+      output_mkl_shape.SetMklLayout(&dst_md);
+      output_mkl_shape.SetElemType(MklDnnType<T>());
+      output_mkl_shape.SetTfLayout(src_mkl_shape.GetDimension(),
+                                   src_mkl_shape.GetSizesAsMklDnnDims(),
+                                   src_mkl_shape.GetTfDataFormat());
+      output_tf_shape.AddDim((dst_pd.get_size() / sizeof(T)) + 1);
```
[the remark this answers is not in the document]
Comment (Srini511) [reply tensorflow/core/kernels/mkl_quantize_op.cc]: Not needed actually. copy paste

File: tensorflow/core/kernels/mkl_quantize_op.cc [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+      min_range = -max_abs;
+      // If it is signed, we try to keep 0.0 being 0 and drop one bucket. For
+      // example, if it is 8 bits, we have the range [-127, 127]. So for input
+      // range of [-x, x], the scale should be 254/(2*x).
+      target_range = static_cast<float>((uint64_t{1} << (num_bits - 1)) - 1);
+    } else {
+      max_range = max_abs;
+      min_range = 0.0;
+      // If it is unsigned and num_bits == 8, the range with 8 bits is [0,
+      // 255].  If the input range is [0, x], then the scale is x/255 instead
```
[the remark this answers is not in the document]
Comment (Srini511) [reply tensorflow/core/kernels/mkl_quantize_op.cc]: Agree

File: tensorflow/core/kernels/mkl_quantize_op.cc [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+    auto reorder_desc = reorder::primitive_desc(src.GetUsrMemPrimDesc(),
+                                                dst.GetUsrMemPrimDesc(), attr);
+    reorder my_reorder = reorder(reorder_desc, primitive::at(*src.GetUsrMem()),
+                                 *dst.GetUsrMem());
+    std::vector<primitive> net{my_reorder};
+    stream(stream::kind::eager).submit(net).wait();
+  }
+
+ private:
+  int mode_;
```
[the remark this answers is not in the document]
Comment (Srini511) [reply tensorflow/core/kernels/mkl_quantize_op.cc]: It is not needed for now, but is meant to be placeholders for future implementation of other types.

File: tensorflow/core/kernels/mkl_quantize_op_test.cc [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+  AddInputFromArray<uint8>(dummy_shape, dummy_tensor);
+  AddInputFromArray<uint8>(dummy_shape, dummy_tensor);
+  AddInputFromArray<uint8>(dummy_shape, dummy_tensor);
+  TF_ASSERT_OK(RunOpKernel());
+  Tensor expected(allocator(), DT_QUINT8, TensorShape({8}));
+  Tensor expected_min(allocator(), DT_FLOAT, TensorShape({}));
+  Tensor expected_max(allocator(), DT_FLOAT, TensorShape({}));
+  // Input element 0.0 should map to 0.
+  // Input element 500.0 is quantized to 255 because max_range = 255.
+  test::FillValues<quint8>(&expected, {0, 1, 1, 2, 127, 255, 255, 2});
```
[the remark this answers is not in the document]
Comment (Srini511) [reply tensorflow/core/kernels/mkl_quantize_op_test.cc]: Yes it is a saturating scaling operation.

**Comment by Srini511** [comment]
Hi @penpornk . Thank you for the review. I have addressed your comments. Please let me know if the changes are okay.

File: tensorflow/core/kernels/mkl_quantize_op.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (penpornk) [reply tensorflow/core/kernels/mkl_quantize_op.cc]: Got it. Thank you!

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the changes and the clarifications! A few more minor changes, please.

**Comment by Srini511** [comment]
Hi @penpornk , I have done the changes above too. Thank you!

**Review by penpornk — APPROVED** [review]
Thank you again for the PR!

**Pull request merged**